File: 51. N-Queens/solveNQueens.py
from typing import List

class Solution:
  def solveNQueens(self, n: int) -> List[List[str]]:
    emptyBoard = [['.'] * n for _ in range(n)]
    diagonal = set()
    antiDiagonal = set()
    cols = set()
    res = []
    
    def creatBoard(board):
      res = []
      for row in range(n):
        res.append(''.join(board[row]))
      
      return res
    
    def backtracking(row, diagonal, antiDiagonal, cols, board):
      if row == n:
        res.append(creatBoard(board))
        return 
      
      for col in range(n):
        currDiagonal = row - col
        currAntiDiagonal = row + col
        
        if (col in cols or
          currDiagonal in diagonal or
          currAntiDiagonal in antiDiagonal):
          continue
        
        board[row][col] = 'Q'
        diagonal.add(currDiagonal)
        antiDiagonal.add(currAntiDiagonal)
        cols.add(col)
        
        backtracking(row + 1, diagonal, antiDiagonal, cols, board)
        
        board[row][col] = '.'
        diagonal.remove(currDiagonal)
        antiDiagonal.remove(currAntiDiagonal)
        cols.remove(col)

    backtracking(0, diagonal, antiDiagonal, cols, emptyBoard)
    return res
  

n = 4 
solution = Solution()
print('solveNQueens: ', solution.solveNQueens(n))

File: 51. N-Queens/solveNQueens.js
/**
 * @param {number} n
 * @return {string[][]}
 */
const solveNQueens = (n) => {
  const res = [];
  const board = Array(n).fill(null).map(() => Array(n).fill('.'));
  const diagonal = new Set();
  const antiDiagonal = new Set();
  const cols = new Set();

  const backtracking = (row) => {
    if (row === n) {
      res.push(board.map(row => row.join('')));
      return;
    }

    for (let col = 0; col < n; col++) {
      const currDiagonal = row - col;
      const currAntiDiagonal = row + col;

      if (cols.has(col) || diagonal.has(currDiagonal) || antiDiagonal.has(currAntiDiagonal)) {
        continue;
      }

      board[row][col] = 'Q';
      cols.add(col);
      diagonal.add(currDiagonal);
      antiDiagonal.add(currAntiDiagonal);

      backtracking(row + 1);

      board[row][col] = '.';
      cols.delete(col);
      diagonal.delete(currDiagonal);
      antiDiagonal.delete(currAntiDiagonal);
    }
  };

  backtracking(0);
  return res;
};


const n = 4;
console.log('solveNQueens: ', solveNQueens(n));
